cpp/src/buffer/buffer.cpp

```diff
@@ -18,20 +20,89 @@ template <typename T>
 }
 }  // namespace
 
+Buffer::Event::Event(rmm::cuda_stream_view stream) {
+    RAPIDSMPF_CUDA_TRY(cudaEventCreateWithFlags(&event_, cudaEventDisableTiming));
+    RAPIDSMPF_CUDA_TRY(cudaEventRecord(event_, stream));
+}
+
+Buffer::Event::~Event() {
+    // First mark as destroying to prevent new readers
+    destroying_.store(true, std::memory_order_release);
+
+    // Wait for any pending is_ready() calls to complete
+    while (active_readers_.load(std::memory_order_acquire) > 0) {
+        std::this_thread::yield();
+    }
+
+    // Now safe to destroy as no readers can start and no readers are active
+    std::lock_guard<std::mutex> lock(mutex_);
+    // TODO: if we're being destroyed, warn the user if the event is
+    // not completed.
+    cudaEventDestroy(event_);
+}
+
+[[nodiscard]] bool Buffer::Event::is_ready() {
+    // First register as an active reader
+    active_readers_.fetch_add(1, std::memory_order_acquire);
+
+    auto exit_if_destroying = [&]() -> std::optional<bool> {
+        if (destroying_.load(std::memory_order_acquire)) {
+            bool done = done_.load(std::memory_order_relaxed);
+            active_readers_.fetch_sub(1, std::memory_order_release);
+            return done;
+        }
+        return std::nullopt;
+    };
+
+    // Early exit if destroying
+    if (auto destroying = exit_if_destroying()) {
+        return *destroying;
+    }
+
+    // Check if we've already determined it's done
+    if (done_.load(std::memory_order_relaxed)) {
+        active_readers_.fetch_sub(1, std::memory_order_release);
+        return true;
+    }
+
+    std::lock_guard<std::mutex> lock(mutex_);
+    // Recheck destroying_ under lock
+    if (auto destroying = exit_if_destroying()) {
+        return *destroying;
+    }
+
+    bool done = done_.load(std::memory_order_relaxed);
+    if (!done) {
+        bool is_complete = (cudaEventQuery(event_) == cudaSuccess);
+        done_.store(is_complete, std::memory_order_release);
+        done = is_complete;
+    }
+
+    active_readers_.fetch_sub(1, std::memory_order_release);
+    return done;
+}
```

Comment:
I think if we rely on the `shared_ptr` storage of `Event`s for thread safety then we don't need the fancy dtor with atomics and mutexes, and we don't need all of this logic in `is_ready` (and can just do the core check of `done_` and `cudaEventQuery`).

Or do we need this for safety even though we have thread-safe destruction with `shared_ptr`?

Reply:
We don't, I explicitly asked for your opinion about that exactly in https://github.com/rapidsai/rapidsmpf/pull/225#discussion_r2071401448 . I think we're in agreement and I'll remove it.

Reply:
Done in https://github.com/rapidsai/rapidsmpf/pull/225/commits/17c01e2c0e7e786112edde871f0c6b109517ba05 .